File: 2024/typescript/day02/day02.ts
const getInputLines = async (fileName: string) => {
  const file = Bun.file(fileName);
  const fileText = await file.text();
  return fileText.split("\n");
};

const getReportsFromLines = (lines: string[]) => {
  return lines.map((line) => line.split(" ").map(Number));
};

const pairwiseDiffs = (report: number[]) => {
  const diffs: number[] = [];
  for (let i = 1; i < report.length; i++) {
    diffs.push(report[i] - report[i - 1]);
  }
  return diffs;
};

const isSafe = (reportDiffs: number[]) => {
  const countNegatives = reportDiffs.filter((v) => v < 0).length;
  if (countNegatives > 0 && countNegatives !== reportDiffs.length) return false;
  const countOutOfBounds = reportDiffs.filter(
    (v) => Math.abs(v) < 1 || Math.abs(v) > 3
  ).length;
  if (countOutOfBounds > 0) return false;
  return true;
};

const part1 = (lines: string[]) => {
  const reports = getReportsFromLines(lines);
  const diffs = reports.map(pairwiseDiffs);
  const safe = diffs.filter((d) => isSafe(d));
  return safe.length;
};

const part2 = (lines: string[]) => {
  let countSafe = 0;
  const reports = getReportsFromLines(lines);
  for (const report of reports) {
    const reportDiffs = pairwiseDiffs(report);
    if (isSafe(reportDiffs)) {
      countSafe++;
    } else {
      for (let i = 0; i < report.length; i++) {
        const updatedReport = report.toSpliced(i, 1);
        const updatedDiffs = pairwiseDiffs(updatedReport);
        if (isSafe(updatedDiffs)) {
          countSafe++;
          break;
        }
      }
    }
  }
  return countSafe;
};

const lines = await getInputLines("day02/day02.input");
console.log("Part 1: ", part1(lines));
console.log("Part 2: ", part2(lines));

File: 2024/day02/day02.py
from itertools import pairwise

with open("day02.input", "r") as f: 
  lines = [list(map(int, l.split())) for l in f.readlines()]

'''
Check that all vals share the same sign (positive or negative)
'''
def same_sign(line_diffs: list[int]):
  count_neg = sum([1 for v in line_diffs if v < 0])
  return count_neg == len(line_diffs) or count_neg == 0

'''
Check that all vals increase or decrease by safe amounts (between 1 and 3)
'''
def safe_change(line_diffs: list[int]):
  safe_changes = len([1 for v in line_diffs if abs(v) >= 1 and abs(v) <= 3])
  return safe_changes == len(line_diffs)

def part1(lines: list[list[int]]):
  safe = 0
  for line in lines:
    diffs = [a - b for (a, b) in pairwise(line)]
    if same_sign(diffs) and safe_change(diffs):
      safe += 1
  return safe

def part2(lines: list[list[int]]):
  safe = 0
  for line in lines:
    diffs = [a - b for (a, b) in pairwise(line)]
    if same_sign(diffs) and safe_change(diffs):
      safe += 1
    else:
      for i in range(len(line)):
        new_line = [v for idx, v in enumerate(line) if idx != i]
        new_diffs = [a - b for (a, b) in pairwise(new_line)]
        if same_sign(new_diffs) and safe_change(new_diffs):
          safe += 1
          break
  return safe
        
print('Part 1: {}'.format(part1(lines)))
print('Part 2: {}'.format(part2(lines)))